zkvm/src/vm.rs

```diff
@@ -51,6 +80,96 @@ pub struct Tx {
     pub proof: R1CSProof,
 }
 
+impl Tx {
+    fn encode(&self, buf: &mut Vec<u8>) {
+        self.header.encode(buf);
+        encoding::write_size(self.program.len(), buf);
+        buf.extend(&self.program);
+        buf.extend_from_slice(&self.signature.to_bytes());
+        buf.extend_from_slice(&self.proof.to_bytes());
+    }
+
+    fn decode<'a>(reader: &mut SliceReader<'a>) -> Result<Tx, VMError> {
+        let (tx, _) = reader.slice(|r| {
+            let header = TxHeader::decode(r)?;
+            let prog_len = r.read_size()?;
+            let program = r.read_bytes(prog_len)?.to_vec();
+
+            let mut sig_buf = [0u8; 64];
+            sig_buf.copy_from_slice(r.read_bytes(64)?);
+            let signature = Signature::from_bytes(sig_buf)?;
+            let proof =
+                R1CSProof::from_bytes(r.read_bytes(r.len())?).map_err(|_| VMError::FormatError)?;
+            Ok(Tx {
+                header,
+                program,
+                signature,
+                proof,
+            })
+        })?;
+        Ok(tx)
+    }
+
+    /// Serializes the tx into a byte array.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(self.serialized_size());
+        self.encode(&mut buf);
+        buf
+    }
+
+    /// Returns the size in bytes required to serialize the `Tx`.
+    pub fn serialized_size(&self) -> usize {
+        // header is 8 bytes * 3 fields = 24 bytes
+        // program length is 4 bytes
+        // program is self.program.len() bytes
+        // signature is 64 bytes
+        // proof is 14*32 + the ipp bytes
+        self.header.serialized_size() + 4 + self.program.len() + 64 + self.proof.serialized_size()
+    }
+
+    /// Deserializes the tx from a byte slice.
+    ///
+    /// Returns an error if the byte slice cannot be parsed into a `Tx`.
+    pub fn from_bytes(slice: &[u8]) -> Result<Tx, VMError> {
+        let tx = SliceReader::parse(slice, |r| Self::decode(r))?;
```

Comment:
no need for extra variables - the result of SliceReader::parse is of the same type as we need.
```suggestion
pub fn from_bytes(slice: &[u8]) -> Result<...> {
    SliceReader::parse(slice, |r| Self::decode(r))
}
```